src/layer/decorators.js

```diff
@@ -0,0 +1,143 @@
+    };
+
+    parent.element.appendChild(this.canvas);
+
+};
+
+(function () {
+    oop.implement(this, EventEmitter);
+
+    this.$updateDecorators = function (config) {
+        var colors = (this.renderer.theme.isDark === true) ? this.colors.dark : this.colors.light;
+        if (config) {
+            this.maxHeight = config.maxHeight;
+            this.lineHeight = config.lineHeight;
+            this.canvasHeight = config.height;
+            var allLineHeight = (config.lastRow + 1) * this.lineHeight;
+            if (allLineHeight < this.canvasHeight) {
+                this.heightRatio = 1;
+            } else {
+                this.heightRatio = this.canvasHeight / this.maxHeight;
+            }
+        }
+        var ctx = this.canvas.getContext("2d");
+
+        function compare(a, b) {
+            if (a.typeNumber < b.typeNumber) return -1;
+            if (a.typeNumber > b.typeNumber) return 1;
+            return 0;
+        }
+
+        var annotations = this.renderer.session.$annotations;
+        ctx.clearRect(0, 0, this.canvas.width, this.canvas.height);
+        if (annotations) {
+            annotations.forEach(function (item) {
+                switch (item.type) {
+                    case "info":
+                        item.typeNumber = 1;
+                        break;
+                    case "warning":
+                        item.typeNumber = 2;
+                        break;
+                    case "error":
+                        item.typeNumber = 3;
+                        break;
+                }
+            });
+            annotations = annotations.sort(compare);
+            var foldData = this.renderer.session.$foldData;
+
+            for (let i = 0; i < annotations.length; i++) {
+                let currentRow = annotations[i].row;
+                let compensateFold = this.compensateFoldRows(currentRow, foldData);
+                let currentY = Math.round((currentRow - compensateFold) * this.lineHeight * this.heightRatio);
+                let y1 = Math.round(((currentRow - compensateFold) * this.lineHeight * this.heightRatio));
+                let y2 = Math.round((((currentRow - compensateFold) * this.lineHeight + this.lineHeight) * this.heightRatio));
+                const height = y2 - y1;
+                if (height < this.minDecorationHeight) {
+                    let yCenter = ((y1 + y2) / 2) | 0;
+                    if (yCenter < this.halfMinDecorationHeight) {
+                        yCenter = this.halfMinDecorationHeight;
+                    }
+                    else if (yCenter + this.halfMinDecorationHeight > this.canvasHeight) {
+                        yCenter = this.canvasHeight - this.halfMinDecorationHeight;
+                    }
+                    y1 = Math.round(yCenter - this.halfMinDecorationHeight);
+                    y2 = Math.round(yCenter + this.halfMinDecorationHeight);
+                }
+
+                switch (annotations[i].type) {
```

Comment:
can we instead simplify this to

```
ctx.fillStyle = colors[annotations[i].type] || null

```

Reply:
Yes, that's definitely better, corrected!